File: services/shared/messages.py
"""WebSocket message definitions for client-backend protocol.

All messages are Pydantic models that serialize to JSON. Frames use a
two-message pattern: JSON metadata followed by binary JPEG bytes.
"""

from __future__ import annotations

import json
import time

from pydantic import BaseModel, Field

from shared.constants import PROTOCOL_VERSION, MessageType


class BaseMessage(BaseModel):
    """Base for all JSON messages."""

    type: str
    timestamp: float = Field(default_factory=time.time)
    protocol_version: str = PROTOCOL_VERSION


# --- Client -> Backend ---


class FrameMetadata(BaseMessage):
    """Metadata sent before a binary JPEG frame."""

    type: str = MessageType.FRAME
    sequence: int
    width: int
    height: int
    jpeg_quality: int
    capture_timestamp: float
    sent_timestamp: float | None = None
    byte_length: int


class H264FrameMetadata(BaseMessage):
    """Metadata sent before a binary H.264 Access Unit."""

    type: str = MessageType.H264_FRAME
    sequence: int
    is_keyframe: bool = False
    capture_timestamp: float
    sent_timestamp: float | None = None
    byte_length: int


class AudioChunk(BaseMessage):
    """Metadata sent before a binary PCM audio chunk."""

    type: str = MessageType.AUDIO_CHUNK
    sequence: int
    sample_rate: int
    channels: int
    sample_format: str = "s16le"
    capture_timestamp: float
    sent_timestamp: float | None = None
    byte_length: int


class ClientStatus(BaseMessage):
    """Periodic status update from client."""

    type: str = MessageType.CLIENT_STATUS
    airplay_running: bool = False
    airplay_pid: int | None = None
    esp32_reachable: bool = False
    esp32_ble_connected: bool | None = None
    frames_captured: int = 0
    frames_sent: int = 0
    capture_source: str = "airplay"
    source_running: bool = False
    control_channel_connected: bool = False
    transport_mode: str = "websocket"
    transport_connected: bool = False
    commands_sent: int = 0
    commands_acked: int = 0
    last_command_rtt_ms: float | None = None
    audio_enabled: bool = False
    audio_source: str | None = None
    audio_chunks_captured: int = 0
    audio_chunks_sent: int = 0
    uptime_seconds: float = 0.0
    # SRT transport stats (populated when transport_mode="srt")
    srt_rtt_ms: float | None = None
    srt_bandwidth_kbps: float | None = None
    srt_packet_loss_pct: float | None = None


# --- Backend -> Client ---


class HIDCommandMessage(BaseMessage):
    """HID command to be forwarded to ESP32."""

    type: str = MessageType.HID_COMMAND
    action: str
    params: dict[str, int | float] = {}
    request_id: str | None = None
    command_id: str | None = None
    command_sequence: int | None = None
    dispatched_at_backend: float | None = None


class CommandAck(BaseMessage):
    """Ack and timing from client after forwarding a command."""

    type: str = MessageType.COMMAND_ACK
    command_id: str
    command_sequence: int | None = None
    received_at_client: float
    forwarded_at_client: float | None = None
    completed_at_client: float
    success: bool = True
    error: str | None = None


class ConfigUpdate(BaseMessage):
    """Backend can adjust client capture parameters dynamically."""

    type: str = MessageType.CONFIG_UPDATE
    jpeg_quality: int | None = None
    max_dimension: int | None = None
    frame_interval_ms: int | None = None


class LocationUpdateMessage(BaseMessage):
    """GPS coordinates to send to the iTools dongle via iOS app."""

    type: str = MessageType.LOCATION_UPDATE
    latitude: float
    longitude: float
    altitude: float = 10.0
    speed_knots: float = 0.0
    heading: float = 0.0


# --- Bidirectional ---


class HeartbeatPing(BaseMessage):
    type: str = MessageType.PING


class HeartbeatPong(BaseMessage):
    type: str = MessageType.PONG


class ErrorMessage(BaseMessage):
    type: str = MessageType.ERROR
    code: str
    detail: str


# --- Message Dispatcher ---

_TYPE_MAP: dict[str, type[BaseMessage]] = {
    MessageType.FRAME: FrameMetadata,
    MessageType.H264_FRAME: H264FrameMetadata,
    MessageType.AUDIO_CHUNK: AudioChunk,
    MessageType.CLIENT_STATUS: ClientStatus,
    MessageType.HID_COMMAND: HIDCommandMessage,
    MessageType.COMMAND_ACK: CommandAck,
    MessageType.CONFIG_UPDATE: ConfigUpdate,
    MessageType.LOCATION_UPDATE: LocationUpdateMessage,
    MessageType.PING: HeartbeatPing,
    MessageType.PONG: HeartbeatPong,
    MessageType.ERROR: ErrorMessage,
}


def parse_message(raw_json: str) -> BaseMessage:
    """Parse a JSON string into the correct message subtype (single parse)."""
    data = json.loads(raw_json)
    msg_type = data.get("type")
    model_class = _TYPE_MAP.get(msg_type, BaseMessage)
    return model_class.model_validate(data)

File: services/ios-app/ChromaCatchController/ChromaCatchController/Protocol/Messages.swift
import Foundation

// MARK: - Protocol Constants

enum MessageType {
    static let frame = "frame"
    static let h264Frame = "h264_frame"
    static let audioChunk = "audio_chunk"
    static let clientStatus = "client_status"
    static let hidCommand = "hid_command"
    static let commandAck = "command_ack"
    static let configUpdate = "config_update"
    static let locationUpdate = "location_update"
    static let error = "error"
    static let ping = "ping"
    static let pong = "pong"
}

// MARK: - Incoming Messages (Backend -> iOS App)

struct LocationUpdateMessage: Codable {
    let type: String
    let latitude: Double
    let longitude: Double
    let altitude: Double
    let speedKnots: Double
    let heading: Double
    let timestamp: Double
    let protocolVersion: String

    enum CodingKeys: String, CodingKey {
        case type, latitude, longitude, altitude, timestamp, heading
        case speedKnots = "speed_knots"
        case protocolVersion = "protocol_version"
    }
}

struct HIDCommandMessage: Codable {
    let type: String
    let action: String
    let params: [String: Double]?
    let commandId: String?
    let commandSequence: Int?
    let dispatchedAtBackend: Double?
    let timestamp: Double

    enum CodingKeys: String, CodingKey {
        case type, action, params, timestamp
        case commandId = "command_id"
        case commandSequence = "command_sequence"
        case dispatchedAtBackend = "dispatched_at_backend"
    }
}

struct HeartbeatPing: Codable {
    let type: String
    let timestamp: Double
}

// MARK: - Outgoing Messages (iOS App -> Backend)

/// Mirrors Python CommandAck — sent after forwarding a HID command to ESP32.
struct CommandAckMessage: Codable {
    let type: String
    let commandId: String
    let commandSequence: Int?
    let receivedAtClient: Double
    let forwardedAtClient: Double?
    let completedAtClient: Double
    let success: Bool
    let error: String?
    let timestamp: Double
    let protocolVersion: String

    enum CodingKeys: String, CodingKey {
        case type, success, error, timestamp
        case commandId = "command_id"
        case commandSequence = "command_sequence"
        case receivedAtClient = "received_at_client"
        case forwardedAtClient = "forwarded_at_client"
        case completedAtClient = "completed_at_client"
        case protocolVersion = "protocol_version"
    }

    init(commandId: String, commandSequence: Int?, receivedAt: Double,
         forwardedAt: Double?, completedAt: Double, success: Bool, error: String? = nil) {
        self.type = MessageType.commandAck
        self.commandId = commandId
        self.commandSequence = commandSequence
        self.receivedAtClient = receivedAt
        self.forwardedAtClient = forwardedAt
        self.completedAtClient = completedAt
        self.success = success
        self.error = error
        self.timestamp = Date().timeIntervalSince1970
        self.protocolVersion = "1.0"
    }
}

/// H.264 frame metadata — sent before binary H.264 Access Unit data.
/// Mirrors Python H264FrameMetadata.
struct H264FrameMetadata: Codable {
    let type: String
    let sequence: Int
    let isKeyframe: Bool
    let captureTimestamp: Double
    let sentTimestamp: Double?
    let byteLength: Int
    let timestamp: Double
    let protocolVersion: String

    enum CodingKeys: String, CodingKey {
        case type, sequence, timestamp
        case isKeyframe = "is_keyframe"
        case captureTimestamp = "capture_timestamp"
        case sentTimestamp = "sent_timestamp"
        case byteLength = "byte_length"
        case protocolVersion = "protocol_version"
    }

    init(sequence: Int, isKeyframe: Bool, captureTimestamp: Double, byteLength: Int) {
        self.type = MessageType.h264Frame
        self.sequence = sequence
        self.isKeyframe = isKeyframe
        self.captureTimestamp = captureTimestamp
        self.sentTimestamp = Date().timeIntervalSince1970
        self.byteLength = byteLength
        self.timestamp = Date().timeIntervalSince1970
        self.protocolVersion = "1.0"
    }
}

/// Audio chunk metadata — sent before binary audio data.
/// Mirrors Python AudioChunk.
struct AudioChunkMetadata: Codable {
    let type: String
    let sequence: Int
    let sampleRate: Int
    let channels: Int
    let sampleFormat: String
    let captureTimestamp: Double
    let sentTimestamp: Double?
    let byteLength: Int
    let timestamp: Double
    let protocolVersion: String

    enum CodingKeys: String, CodingKey {
        case type, sequence, channels, timestamp
        case sampleRate = "sample_rate"
        case sampleFormat = "sample_format"
        case captureTimestamp = "capture_timestamp"
        case sentTimestamp = "sent_timestamp"
        case byteLength = "byte_length"
        case protocolVersion = "protocol_version"
    }

    init(sequence: Int, sampleRate: Int, channels: Int,
         sampleFormat: String = "s16le", captureTimestamp: Double, byteLength: Int) {
        self.type = MessageType.audioChunk
        self.sequence = sequence
        self.sampleRate = sampleRate
        self.channels = channels
        self.sampleFormat = sampleFormat
        self.captureTimestamp = captureTimestamp
        self.sentTimestamp = Date().timeIntervalSince1970
        self.byteLength = byteLength
        self.timestamp = Date().timeIntervalSince1970
        self.protocolVersion = "1.0"
    }
}

/// Client status — mirrors Python ClientStatus with iOS-specific extras.
/// Pydantic ignores unknown fields, so the iOS extras are safe to include.
struct ClientStatus: Codable {
    let type: String
    // Standard fields (matching Python ClientStatus)
    let airplayRunning: Bool
    let esp32Reachable: Bool
    let esp32BleConnected: Bool?
    let framesCaptured: Int
    let framesSent: Int
    let captureSource: String
    let sourceRunning: Bool
    let controlChannelConnected: Bool
    let transportMode: String
    let transportConnected: Bool
    let commandsSent: Int
    let commandsAcked: Int
    let lastCommandRttMs: Double?
    let audioEnabled: Bool
    let audioSource: String?
    let audioChunksCaptured: Int
    let audioChunksSent: Int
    let uptimeSeconds: Double
    // iOS-specific extras (Pydantic ignores unknown fields)
    let eaConnected: Bool
    let bleConnected: Bool
    let dongleForwarding: Bool
    let currentLatitude: Double?
    let currentLongitude: Double?
    let timestamp: Double
    let protocolVersion: String

    enum CodingKeys: String, CodingKey {
        case type, timestamp
        case airplayRunning = "airplay_running"
        case esp32Reachable = "esp32_reachable"
        case esp32BleConnected = "esp32_ble_connected"
        case framesCaptured = "frames_captured"
        case framesSent = "frames_sent"
        case captureSource = "capture_source"
        case sourceRunning = "source_running"
        case controlChannelConnected = "control_channel_connected"
        case transportMode = "transport_mode"
        case transportConnected = "transport_connected"
        case commandsSent = "commands_sent"
        case commandsAcked = "commands_acked"
        case lastCommandRttMs = "last_command_rtt_ms"
        case audioEnabled = "audio_enabled"
        case audioSource = "audio_source"
        case audioChunksCaptured = "audio_chunks_captured"
        case audioChunksSent = "audio_chunks_sent"
        case uptimeSeconds = "uptime_seconds"
        case eaConnected = "ea_connected"
        case bleConnected = "ble_connected"
        case dongleForwarding = "dongle_forwarding"
        case currentLatitude = "current_latitude"
        case currentLongitude = "current_longitude"
        case protocolVersion = "protocol_version"
    }

    init(eaConnected: Bool, bleConnected: Bool, dongleForwarding: Bool,
         esp32Reachable: Bool = false, controlChannelConnected: Bool = false,
         transportMode: String = "ios-replaykit", transportConnected: Bool = false,
         framesCaptured: Int = 0, framesSent: Int = 0,
         commandsSent: Int = 0, commandsAcked: Int = 0,
         lastCommandRttMs: Double? = nil,
         audioChunksCaptured: Int = 0, audioChunksSent: Int = 0,
         currentLatitude: Double? = nil, currentLongitude: Double? = nil,
         uptimeSeconds: Double = 0) {
        self.type = MessageType.clientStatus
        self.airplayRunning = false  // iOS doesn't use AirPlay
        self.esp32Reachable = esp32Reachable
        self.esp32BleConnected = nil
        self.framesCaptured = framesCaptured
        self.framesSent = framesSent
        self.captureSource = "ios-replaykit"
        self.sourceRunning = false
        self.controlChannelConnected = controlChannelConnected
        self.transportMode = transportMode
        self.transportConnected = transportConnected
        self.commandsSent = commandsSent
        self.commandsAcked = commandsAcked
        self.lastCommandRttMs = lastCommandRttMs
        self.audioEnabled = false
        self.audioSource = nil
        self.audioChunksCaptured = audioChunksCaptured
        self.audioChunksSent = audioChunksSent
        self.uptimeSeconds = uptimeSeconds
        self.eaConnected = eaConnected
        self.bleConnected = bleConnected
        self.dongleForwarding = dongleForwarding
        self.currentLatitude = currentLatitude
        self.currentLongitude = currentLongitude
        self.timestamp = Date().timeIntervalSince1970
        self.protocolVersion = "1.0"
    }
}

struct HeartbeatPong: Codable {
    let type: String
    let timestamp: Double

    init() {
        self.type = MessageType.pong
        self.timestamp = Date().timeIntervalSince1970
    }
}

// MARK: - Message Parsing

enum IncomingMessage {
    case locationUpdate(LocationUpdateMessage)
    case hidCommand(HIDCommandMessage)
    case ping(HeartbeatPing)
    case unknown(String)

    static func parse(_ text: String) -> IncomingMessage? {
        guard let data = text.data(using: .utf8) else { return nil }

        // Peek at the type field
        guard let json = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = json["type"] as? String else {
            return nil
        }

        let decoder = JSONDecoder()
        switch type {
        case MessageType.locationUpdate:
            guard let msg = try? decoder.decode(LocationUpdateMessage.self, from: data) else { return nil }
            return .locationUpdate(msg)
        case MessageType.hidCommand:
            guard let msg = try? decoder.decode(HIDCommandMessage.self, from: data) else { return nil }
            return .hidCommand(msg)
        case MessageType.ping:
            guard let msg = try? decoder.decode(HeartbeatPing.self, from: data) else { return nil }
            return .ping(msg)
        default:
            return .unknown(type)
        }
    }
}
